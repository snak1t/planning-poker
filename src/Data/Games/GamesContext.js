import React, { useState, useContext } from 'react';
import { message } from 'antd';
import filter from 'ramda/src/filter';
import append from 'ramda/src/append';
import { useAsyncEffect } from '../../utils/hooks/useAsyncEffect';
import { AuthContext } from '../Auth/AuthContext';
import { ApiClient } from '../../utils/api-client';
export const GamesContext = React.createContext();

export function GamesProvider({ initialGames = [], children, readyToFetch }) {
    const [games, setGames] = useState(initialGames);
    const { user } = useContext(AuthContext);
    useAsyncEffect(
        async () => {
            if (!readyToFetch) {
                return;
            }
            try {
                const { data } = await ApiClient.get(`/api/game/user/${user.info.email}`);
                return setGames(data.games);
            } catch (error) {
                console.error(error);
            }
        },
        [readyToFetch],
    );

    const removeGame = async gameId => {
        try {
            const { data } = await ApiClient.delete(`/api/game/`, { params: { gameId } });
            setGames(filter(game => game.id !== data.id));
        } catch (error) {
            message.error(error.message);
        }
    };
    const addGame = async formData => {
        try {
            const { data } = await ApiClient.post('/api/game', { ...formData, user: user.info.email });
            setGames(append(data.game));
        } catch (error) {
            message.error(error.message);
        }
    };

    const updateGame = updatedGame => {
        setGames(prevGames => {
            const isGamePresent = prevGames.find(game => game.id === updatedGame.id);
            if (isGamePresent) {
                return prevGames.map(game => (game.id === updatedGame.id ? updatedGame : game));
            } else {
                return prevGames.concat(updatedGame);
            }
        });
    };
    return <GamesContext.Provider value={{ games, removeGame, addGame, updateGame }}>{children}</GamesContext.Provider>;
}

export const useCurrentGame = gameId => {
    const { games } = useContext(GamesContext);
    const game = games.find(game => game.id.toString() === gameId);
    return game;
};
