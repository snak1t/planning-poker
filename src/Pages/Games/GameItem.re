open Ant;

let component = ReasonReact.statelessComponent("GameItem");

let make =
    (
      ~title: string,
      ~description: string,
      ~id: string,
      ~onNavigateToTask: string => unit,
      ~onDeleteGame: string => unit,
      _children,
    ) => {
  let onPlay = _e => onNavigateToTask(id);
  let onDelete = _e => onDeleteGame(id);
  let renderDescription = (desc: string) =>
    ReasonReact.string(
      switch (desc) {
      | "" => "There is no description"
      | _ => desc
      },
    );
  {
    ...component,
    render: _self =>
      <Card
        style={ReactDOMRe.Style.make(~margin="0.4rem", ())}
        title
        actions=[|
          <Icon _type=Icon.PlayCircle onClick=onPlay />,
          <Icon _type=Icon.Delete onClick=onDelete />,
        |]>
        {renderDescription(description)}
      </Card>,
  };
};

[@bs.deriving abstract]
type jsProps = {
  title: string,
  description: string,
  onNavigateToTask: string => unit,
  onDeleteGame: string => unit,
  _id: string,
};

let default =
  ReasonReact.wrapReasonForJs(~component, jsProps =>
    make(
      ~title=jsProps##title,
      ~description=jsProps##description,
      ~id=jsProps##_id,
      ~onNavigateToTask=jsProps##onNavigateToTask,
      ~onDeleteGame=jsProps##onDeleteGame,
      [||],
    )
  );