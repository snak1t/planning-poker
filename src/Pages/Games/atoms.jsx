import styled from 'styled-components';

export const GamesHeader = styled.header`
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 0.4rem 1rem;
    h1 {
        font-size: 1.2rem;
        line-height: 1.4rem;
    }
`;

export const GamesGrid = styled.section`
    padding: 2rem;
    display: grid;
    grid-template-columns: repeat(5, 1fr);
    grid-gap: 1rem;
    min-height: 0;
    min-width: 0;
    @media screen and (max-width: 400px) {
        padding: 0.2rem;
        grid-template-columns: 1fr;
    }
`;

export const GameGridItem = styled.article`
    min-width: 0;
`;
