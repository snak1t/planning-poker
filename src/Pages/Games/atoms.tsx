import styled from 'styled-components';

export const GamesHeader = styled.header`
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 0.4rem 1rem;
    h1 {
        font-size: 1.2rem;
        line-height: 1.4rem;
    }
`;

export const GamesWrapper = styled.section`
    padding: 2rem;
`;
