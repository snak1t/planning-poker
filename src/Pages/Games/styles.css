.Games-container {
    flex: 1;
}

.Games-list {
    display: flex;
    flex-wrap: wrap;
}

.Games-item {
    flex-basis: 300px;
    margin: 10px;
    min-height: 200px;
    display: flex;
    flex-direction: column;
    box-sizing: border-box;
    background-color: #fff;
    padding: 10px;
    box-shadow: 0 0 6px 1px rgba(0, 0, 0, 0.38);
}

.Games-gameHeader {
    line-height: 1.6rem;
    font-size: 1.2rem;
    text-align: center;
    font-weight: bold;
}

.Games-gameDescription {
    flex-grow: 1;
    display: flex;
    align-items: center;
    justify-content: center;
    text-align: center;
}

.Games-gameFooter {
    margin: 10px 0px;
    display: flex;
    justify-content: space-around;
}
