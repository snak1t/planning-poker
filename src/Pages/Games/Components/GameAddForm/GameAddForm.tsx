import React, { useContext } from 'react';
import { Modal, Input, Icon, Radio, Form, message } from 'antd';
import { useTextField } from '../../../../utils/hooks/useTextField';
import { GamesContext } from '../../../../Data/Games/GamesContext';

const auditIcon = <Icon type="audit" style={{ color: 'rgba(0,0,0,.25)' }} />;

type Props = {
    onClose: () => void;
};
export const GameAddForm: React.SFC<Props> = ({ onClose }) => {
    const [description, setDescription] = useTextField('');
    const [title, setTitle] = useTextField('');
    const { addGame } = useContext(GamesContext);
    const onSubmit = () => {
        if (title === '') {
            return message.info('Each game should have a title');
        }
        addGame({ description, title });
        onClose();
    };
    return (
        <Modal visible={true} title="Create new game" onCancel={onClose} onOk={onSubmit}>
            <Form.Item>
                <Input
                    prefix={auditIcon}
                    id="title"
                    name="title"
                    placeholder="Game Title"
                    data-testid="game-title-input"
                    value={title}
                    onChange={setTitle}
                />
            </Form.Item>
            <Form.Item>
                <Input.TextArea
                    id="description"
                    name="description"
                    placeholder="Optional game description"
                    data-testid="game-description-input"
                    value={description}
                    onChange={setDescription}
                />
            </Form.Item>
            <Form.Item>
                <Radio.Group defaultValue="fibonacci" buttonStyle="solid">
                    <Radio.Button value="fibonacci">Fibonacci</Radio.Button>
                    <Radio.Button value="sizes">Sizes</Radio.Button>
                </Radio.Group>
            </Form.Item>
        </Modal>
    );
};
