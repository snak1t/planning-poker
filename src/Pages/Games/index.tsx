import React, { useState, useContext } from 'react';
import { not } from 'ramda';
import { Button } from 'antd';

import { GamesList } from './Components/List';
import { GameForm } from './Components/Form';
import * as Atoms from './atoms';
import { QuickGame } from '../../components/QuickGame';
import { AuthContext } from '../../Data/Auth/AuthContext';
import { RouteComponentProps } from 'react-router';
import { FloatButtonsContainer } from '../../components/FloatButtonsContainer/FloatButtonsContainer';

const GamesContainer: React.SFC<RouteComponentProps> = () => {
    const { logout } = useContext(AuthContext);
    const [modalWindow, setModalWindowVisibility] = useState(false);
    const toggleModalWindow = () => setModalWindowVisibility(not);

    return (
        <>
            {modalWindow ? <GameForm onClose={toggleModalWindow} /> : null}
            <Atoms.GamesHeader>
                <h2>Your games</h2>
                <QuickGame />
                <Button onClick={logout}>Sign out</Button>
            </Atoms.GamesHeader>
            <GamesList />
            <FloatButtonsContainer>
                <Button onClick={toggleModalWindow} type="primary" shape="circle" icon="plus" size="large" />
            </FloatButtonsContainer>
        </>
    );
};

export default GamesContainer;
