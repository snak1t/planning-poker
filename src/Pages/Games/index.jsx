import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import evolve from 'ramda/src/evolve'
import not from 'ramda/src/not'

import './styles.css'
import GamesList from './Components/List'
import GameForm from './Components/Form'

export class GamesConatainer extends React.Component {
  state = {
    modal: false
  }

  navigateToGame = id =>
    this.props.history.replace(`/game/${this.props.userLogin}/${id}`)

  toggleModalWindow = () => this.setState(evolve({ modal: not }))

  render() {
    return (
      <div className="Games-container">
        {this.state.modal
          ? <GameForm onClose={this.toggleModalWindow} />
          : null}
        <div className="Games-title">
          <h1>Your games</h1>
        </div>
        <button onClick={this.toggleModalWindow}>Add Form</button>
        <GamesList onNavigateToTask={this.navigateToGame} />
      </div>
    )
  }
}

GamesConatainer.propTypes = {
  history: PropTypes.object.isRequired,
  userLogin: PropTypes.string
}

const mapStateToProps = state => ({
  userLogin: state.user.login
})

const mapDispatchToProps = {}

export default connect(mapStateToProps, mapDispatchToProps)(GamesConatainer)
