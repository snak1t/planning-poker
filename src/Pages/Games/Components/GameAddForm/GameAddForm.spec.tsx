import React from 'react';
import { render, cleanup, fireEvent, Matcher } from 'react-testing-library';
import { GameAddForm } from './GameAddForm';
import { GamesContext } from '../../../../Data/Games/GamesContext';
import message from 'antd/lib/message';
jest.mock('antd/lib/message');

describe('<GameAddForm /> ', () => {
    let onClose: jest.Mock<{}>;
    let removeGame: jest.Mock<{}>;
    let addGame: jest.Mock<{}>;
    let updateGame: jest.Mock<{}>;
    let gamesProvidesMock: {
        games: [];
        addGame: typeof addGame;
        updateGame: typeof updateGame;
        removeGame: typeof removeGame;
    };
    beforeEach(() => {
        onClose = jest.fn();
        removeGame = jest.fn();
        addGame = jest.fn();
        updateGame = jest.fn();
        gamesProvidesMock = {
            games: [],
            addGame,
            removeGame,
            updateGame,
        };
    });
    afterEach(cleanup);

    const updateFormValues = (
        getFn: (text: Matcher) => HTMLElement,
        values: { description: string; title: string },
    ) => {
        const titleInput = getFn('game-title-input') as HTMLInputElement;
        const descriptionInput = getFn('game-description-input') as HTMLInputElement;

        fireEvent.change(titleInput, {
            target: {
                value: values.title,
            },
        });
        fireEvent.change(descriptionInput, {
            target: {
                value: values.description,
            },
        });
    };

    it('should be rendered', () => {
        const { baseElement } = render(
            <GamesContext.Provider value={gamesProvidesMock}>
                <GameAddForm onClose={onClose} />
            </GamesContext.Provider>,
        );
        expect(baseElement).toMatchSnapshot();
    });

    it('should trigger on close button on closing the window', () => {
        const { getByText } = render(
            <GamesContext.Provider value={gamesProvidesMock}>
                <GameAddForm onClose={onClose} />
            </GamesContext.Provider>,
        );
        const cancelButton = getByText('Cancel').parentElement;
        if (cancelButton) {
            fireEvent.click(cancelButton);
        }
        expect(onClose).toHaveBeenCalled();
    });

    it('should trigger addGame function on adding the game', () => {
        const { getByText, getByTestId } = render(
            <GamesContext.Provider value={gamesProvidesMock}>
                <GameAddForm onClose={onClose} />
            </GamesContext.Provider>,
        );
        const okButton = getByText('OK').parentElement;
        const testedData = {
            title: 'Sprint 12',
            description: 'Some Description',
        };
        updateFormValues(getByTestId, testedData);
        if (okButton) {
            fireEvent.click(okButton);
        }
        expect(addGame).toHaveBeenCalled();
        expect(addGame).toHaveBeenCalledWith(testedData);
    });

    it('should close the window after adding the game', () => {
        const { getByText, getByTestId } = render(
            <GamesContext.Provider value={gamesProvidesMock}>
                <GameAddForm onClose={onClose} />
            </GamesContext.Provider>,
        );
        const testedData = {
            title: 'Sprint 12',
            description: 'Some Description',
        };
        updateFormValues(getByTestId, testedData);
        const okButton = getByText('OK').parentElement;
        if (okButton) {
            fireEvent.click(okButton);
        }
        expect(onClose).toHaveBeenCalled();
    });
    it('should not add game with missing title property', () => {
        const { getByText, getByTestId } = render(
            <GamesContext.Provider value={gamesProvidesMock}>
                <GameAddForm onClose={onClose} />
            </GamesContext.Provider>,
        );
        const okButton = getByText('OK').parentElement;
        const testedData = {
            title: '',
            description: 'Some Description',
        };
        updateFormValues(getByTestId, testedData);
        if (okButton) {
            fireEvent.click(okButton);
        }
        expect(addGame).not.toHaveBeenCalled();
        expect(message.info).toHaveBeenCalledWith('Each game should have a title');
    });
});
