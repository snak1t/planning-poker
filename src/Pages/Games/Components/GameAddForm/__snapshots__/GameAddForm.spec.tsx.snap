// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameAddForm />  should be rendered 1`] = `null`;
